utils/include/detray/simulation/event_generator/random_track_generator.hpp

```diff
@@ -16,47 +16,77 @@
 struct random_numbers {
 
-    random_numbers(random_numbers&& other) : engine(std::move(other.engine)) {}
+    std::seed_seq m_seeds;
+    engine_t m_engine;
 
-    generator_t gen;
-    engine_t engine;
+    /// Random seed for increased entropy. Different random numbers each time
+    DETRAY_HOST
+    random_numbers() : m_seeds{get_random_seeds()}, m_engine{m_seeds} {}
 
-    template <
-        typename T = generator_t,
-        std::enable_if_t<std::is_same_v<T, std::random_device>, bool> = true>
-    random_numbers() : gen{}, engine{gen()} {}
+    /// Deterministic @param seed for reproducible results
+    DETRAY_HOST
+    random_numbers(std::size_t seed) : m_seeds{seed}, m_engine{m_seeds} {}
 
-    template <typename T = generator_t,
-              std::enable_if_t<std::is_same_v<T, std::seed_seq>, bool> = true>
-    random_numbers() : gen{42u}, engine{gen} {}
+    /// Copy constructor
+    DETRAY_HOST
+    random_numbers(random_numbers&& other)
+        : m_engine(std::move(other.m_engine)) {}
 
-    template <typename T = distribution_t,
-              std::enable_if_t<
-                  std::is_same_v<T, std::uniform_real_distribution<scalar_t>>,
-                  bool> = true>
-    DETRAY_HOST auto operator()(const scalar_t min, const scalar_t max) {
-        return distribution_t(min, max)(engine);
+    /// Generate random numbers in a given range
+    DETRAY_HOST auto operator()(const std::array<scalar_t, 2> range = {
+                                    -std::numeric_limits<scalar_t>::max(),
+                                    std::numeric_limits<scalar_t>::max()}) {
+        const scalar_t min{range[0]}, max{range[1]};
+        assert(min <= max);
+
+        // Uniform
+        if constexpr (std::is_same_v<
+                          distribution_t,
+                          std::uniform_real_distribution<scalar_t>>) {
+
+            return distribution_t(min, max)(m_engine);
+
+            // Normal
+        } else if constexpr (std::is_same_v<
+                                 distribution_t,
+                                 std::normal_distribution<scalar_t>>) {
+
+            scalar_t mu{min + 0.5f * (max - min)};
+            return distribution_t(mu, 0.5f / 3.0f * (max - min))(m_engine);
+        }
     }
 
-    template <
-        typename T = distribution_t,
-        std::enable_if_t<std::is_same_v<T, std::normal_distribution<scalar_t>>,
-                         bool> = true>
-    DETRAY_HOST auto operator()(const scalar_t min, const scalar_t max) {
-        scalar_t mu{min + 0.5f * (max - min)};
-        return distribution_t(mu, 0.5f / 3.0f * (max - min))(engine);
+    /// Explicit normal distribution around a @param mean and @param stddev
+    DETRAY_HOST auto normal(const scalar_t mean, const scalar_t stddev) {
+        return std::normal_distribution<scalar_t>(mean, stddev)(m_engine);
+    }
+
+    /// Get proper random seeds for RNG
+    DETRAY_HOST
+    auto get_random_seeds() {
```

Comment:
Can we just set the default seed as 0? (or force user to set the seed in the constructor) I don't see a good point in seeting seeds randomly by default.

Reply:
I am not sure setting all bits to zero is a good seed. This is just a more stable way of seeding, while using a single deterministic seed will give reproducible results

Reply:
Even this `get_random_seeds` will give us the same sequence of seeds for different executions, which is still deterministic

Reply:
That depends on how ```std::random_device``` is implemented on your plattform. On my laptop it does not produce the same random numbers on different executions

Reply:
"std::random_device is a uniformly-distributed integer random number generator that produces non-deterministic random numbers. " - cppreference

It is not guaranteed though, I believe, it depends on the implementation of your plattform. But it should still pick up some entroy from e.g. the OS, so I would be surprised if it delivered the same sequence every time

Reply:
That is because you used the ```std::seed_seq``` type in the template parameters of ```random_numbers```? This set the seed to 42, so that the random number for the unittest were reproducible. Now, I set that explicitely in the unittest and it should result in the same

Reply:
That might explain that